File: apps/server-v1/src/tests/api/routes/test_users.py
from fastapi.testclient import TestClient

from src.crud import UserService
from src.tests.utils.users import create_random_user


def test_create_user(client: TestClient, user_service: UserService) -> None:
    """Test to create a new user"""
    user_data = create_random_user(user_service, False)
    response = client.post("/api/v1/users/", json=user_data.model_dump())

    assert response.status_code == 201
    data = response.json()
    assert data["email"] == user_data.email
    assert data["name"] == user_data.name
    assert data["age"] == user_data.age
    assert data["is_active"] is True


def test_read_user(client: TestClient, user_service: UserService) -> None:
    """Test to retrieve a user by ID"""
    user = create_random_user(user_service, True)
    response = client.get(f"/api/v1/users/{user.id}")

    assert response.status_code == 200
    data = response.json()
    assert data["id"] == user.id
    assert data["email"] == user.email
    assert data["name"] == user.name
    assert data["age"] == user.age
    assert data["is_active"] is True


def test_read_user_not_found(client: TestClient) -> None:
    """Test to retrieve a user that does not exist"""
    response = client.get("/api/v1/users/-1")
    assert response.status_code == 404
    assert response.json()["detail"] == "User not found"


def test_get_users(client: TestClient, user_service: UserService) -> None:
    """Test to retrieve a list of users"""
    # Create at least 5 users
    for _ in range(5):
        create_random_user(user_service, True)

    response = client.get("/api/v1/users/")
    assert response.status_code == 200
    data = response.json()
    assert isinstance(data, list)
    assert len(data) >= 5  # Make sure at least 5 users are returned


def test_update_user(client: TestClient, user_service: UserService) -> None:
    """Test to update an existing user"""
    user = create_random_user(user_service, True)
    update_data = {"name": "Updated Name", "age": 30}

    response = client.put(f"/api/v1/users/{user.id}", json=update_data)
    assert response.status_code == 200
    data = response.json()
    assert data["id"] == user.id
    assert data["name"] == update_data["name"]
    assert data["age"] == update_data["age"]


def test_update_user_not_found(client: TestClient) -> None:
    """Test to update a user that does not exist"""
    update_data = {"name": "Updated Name", "age": 30}
    response = client.put("/api/v1/users/-1", json=update_data)
    assert response.status_code == 404
    assert response.json()["detail"] == "User not found"


def test_delete_user(client: TestClient, user_service: UserService) -> None:
    """Test to delete an existing user"""
    user = create_random_user(user_service, True)

    response = client.delete(f"/api/v1/users/{user.id}")
    assert response.status_code == 204

    # Verify the user is actually deleted
    response = client.get(f"/api/v1/users/{user.id}")
    assert response.status_code == 404


def test_delete_user_not_found(client: TestClient) -> None:
    """Test to delete a user that does not exist"""
    response = client.delete("/api/v1/users/-1")
    assert response.status_code == 404
    assert response.json()["detail"] == "User not found"

File: apps/server-v2/src/tests/api/routes/test_users.rs
use actix_web::http::StatusCode;
use actix_web::{self, App, test};

use crate::api::main::handler;
use crate::models::users::Model as UserModel;
use crate::schemas::users::{UserCreate, UserUpdate};

use crate::tests::utils::api::TestAPIParameters;
use crate::tests::utils::users::create_random_user;
use crate::tests::utils::utils::{random_email, random_int, random_string};

#[actix_web::test]
async fn test_create_user() {
    let api_params = TestAPIParameters::new().await;
    let app = test::init_service(
        App::new()
            .app_data(api_params.app_data.clone())
            .service(handler(api_params.prefix.as_str())),
    )
    .await;

    let user_create = UserCreate {
        email: random_email(),
        name: random_string(32),
        age: Some(random_int(18, 65)),
    };

    let req = test::TestRequest::post()
        .uri(&format!("{}/users/", api_params.prefix))
        .set_json(user_create.clone())
        .to_request();

    let resp = test::call_service(&app, req).await;
    assert_eq!(resp.status(), StatusCode::CREATED);
    let resp: UserModel = test::read_body_json(resp).await;

    assert_eq!(resp.email, user_create.email);
    assert_eq!(resp.name, user_create.name);
    assert_eq!(resp.age, user_create.age);
    assert_eq!(resp.is_active, Some(true));
}

#[actix_web::test]
async fn test_read_user() {
    let api_params = TestAPIParameters::new().await;
    let app = test::init_service(
        App::new()
            .app_data(api_params.app_data.clone())
            .service(handler(api_params.prefix.as_str())),
    )
    .await;

    let user = create_random_user(true, Some(&api_params.db.clone().connection))
        .await
        .as_model()
        .expect("Failed to create user");

    let req = test::TestRequest::get()
        .uri(&format!("{}/users/id/{}", api_params.prefix, user.id))
        .to_request();

    let resp: UserModel = test::call_and_read_body_json(&app, req).await;

    assert_eq!(resp.id, user.id);
    assert_eq!(resp.email, user.email);
    assert_eq!(resp.name, user.name);
    assert_eq!(resp.age, user.age);
    assert_eq!(resp.is_active, Some(true));
}

#[actix_web::test]
async fn test_read_user_not_found() {
    let api_params = TestAPIParameters::new().await;
    let app = test::init_service(
        App::new()
            .app_data(api_params.app_data.clone())
            .service(handler(api_params.prefix.as_str())),
    )
    .await;

    // Assuming 999999 is an ID that does not exist
    let req = test::TestRequest::get()
        .uri(&format!("{}/users/id/999999", api_params.prefix))
        .to_request();

    let resp = test::call_service(&app, req).await;
    assert_eq!(resp.status(), StatusCode::NOT_FOUND);
}

#[actix_web::test]
async fn test_get_users() {
    let api_params = TestAPIParameters::new().await;
    let app = test::init_service(
        App::new()
            .app_data(api_params.app_data.clone())
            .service(handler(api_params.prefix.as_str())),
    )
    .await;

    // Create 5 users directly in DB
    for _ in 0..5 {
        let _ = create_random_user(true, Some(&api_params.db.connection)).await;
    }

    let req = test::TestRequest::get()
        .uri(&format!("{}/users/", api_params.prefix))
        .to_request();

    let resp = test::call_service(&app, req).await;
    assert_eq!(resp.status(), StatusCode::OK);
    let users: Vec<UserModel> = test::read_body_json(resp).await;
    assert!(users.len() >= 5);
}

#[actix_web::test]
async fn test_update_user() {
    let api_params = TestAPIParameters::new().await;
    let app = test::init_service(
        App::new()
            .app_data(api_params.app_data.clone())
            .service(handler(api_params.prefix.as_str())),
    )
    .await;

    let user = create_random_user(true, Some(&api_params.db.connection))
        .await
        .as_model()
        .unwrap();

    let update = UserUpdate {
        name: Some("updated name".to_string()),
        age: Some(30),
        email: None,
        is_active: None,
    };

    let req = test::TestRequest::put()
        .uri(&format!("{}/users/id/{}", api_params.prefix, user.id))
        .set_json(update.clone())
        .to_request();

    let resp = test::call_service(&app, req).await;
    assert_eq!(resp.status(), StatusCode::OK);
    let updated: UserModel = test::read_body_json(resp).await;
    assert_eq!(updated.id, user.id);
    assert_eq!(updated.name, update.name.unwrap());
    assert_eq!(updated.age, update.age);
    // Email unchanged
    assert_eq!(updated.email, user.email);
}

#[actix_web::test]
async fn test_update_user_not_found() {
    let api_params = TestAPIParameters::new().await;
    let app = test::init_service(
        App::new()
            .app_data(api_params.app_data.clone())
            .service(handler(api_params.prefix.as_str())),
    )
    .await;

    let update = UserUpdate {
        name: Some("updated name".to_string()),
        age: Some(30),
        email: None,
        is_active: None,
    };

    let req = test::TestRequest::put()
        .uri(&format!("{}/users/id/999999", api_params.prefix))
        .set_json(update)
        .to_request();

    let resp = test::call_service(&app, req).await;
    assert_eq!(resp.status(), StatusCode::NOT_FOUND);
}

#[actix_web::test]
async fn test_delete_user() {
    let api_params = TestAPIParameters::new().await;
    let app = test::init_service(
        App::new()
            .app_data(api_params.app_data.clone())
            .service(handler(api_params.prefix.as_str())),
    )
    .await;

    let user = create_random_user(true, Some(&api_params.db.connection))
        .await
        .as_model()
        .unwrap();

    let req = test::TestRequest::delete()
        .uri(&format!("{}/users/id/{}", api_params.prefix, user.id))
        .to_request();
    let resp = test::call_service(&app, req).await;
    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    // Verify deleted
    let req = test::TestRequest::get()
        .uri(&format!("{}/users/id/{}", api_params.prefix, user.id))
        .to_request();
    let resp = test::call_service(&app, req).await;
    assert_eq!(resp.status(), StatusCode::NOT_FOUND);
}

#[actix_web::test]
async fn test_delete_user_not_found() {
    let api_params = TestAPIParameters::new().await;
    let app = test::init_service(
        App::new()
            .app_data(api_params.app_data.clone())
            .service(handler(api_params.prefix.as_str())),
    )
    .await;

    let req = test::TestRequest::delete()
        .uri(&format!("{}/users/id/999999", api_params.prefix))
        .to_request();
    let resp = test::call_service(&app, req).await;
    assert_eq!(resp.status(), StatusCode::NOT_FOUND);
}
